services/construction_service.go

```diff
@@ -264,7 +290,7 @@ func (s *ConstructionAPIService) ConstructionMetadata(
 		to = checkTokenContractAddress
 
 		var err *types.Error
-		gasLimit, err = s.calculateGasLimit(ctx, checkFrom, checkTokenContractAddress, input.Data, nil)
+		gasLimit, err = s.calculateGasLimit(ctx, checkFrom, checkTokenContractAddress, input.Data, nil, input.GasLimit)
```

Comment:
Hmm, I think I prefer the pattern where you don't even call this method if input.GasLimit is non-nill (since you've already done the Uint64 conversion for that case above).  You can then drop the added parameter.

Reply:
to be fair i think some of this may have been inherited from the polygon implementation; +1 on your suggested pattern